snapcraft/internal/errors.py

```diff
@@ -41,13 +41,24 @@ def __init__(self, **kwargs):
     def __str__(self):
         return self.fmt.format([], **self.__dict__)
 
+    @property
```

Comment:
All I really care about here is API design: this is designed to be reimplemented by subclasses. I find that to be clear given a function API, making it an attribute makes that relationship less clear and more difficult to use. I'd prefer having this be either a property (function) or a real function.